hie-hare/Haskell/Ide/HaRePlugin.hs

```diff
@@ -312,96 +315,137 @@ makeRefactorResult changedFiles = do
   return $ fold diffs
 
 -- ---------------------------------------------------------------------
+nonExistentCacheErr :: String -> IdeResponse a
+nonExistentCacheErr meth =
+  IdeResponseFail $
+    IdeError PluginError
+             (T.pack $ meth <> ": \"" <> "module not loaded" <> "\"")
+             Null
+
+invalidCursorErr :: String -> IdeResponse a
+invalidCursorErr meth =
+  IdeResponseFail $
+    IdeError PluginError
+             (T.pack $ meth <> ": \"" <> "Invalid cursor position" <> "\"")
+             Null
+-- ---------------------------------------------------------------------
 
-findDefCmd :: TextDocumentPositionParams -> IdeM (IdeResponse Location)
-findDefCmd (TextDocumentPositionParams tdi pos) = do
-    cms <- getCachedModules
-    eitherRes <- runHareCommand' $ findDef cms (tdi ^. J.uri) (unPos pos)
-    case eitherRes of
-      Right x -> return x
-      Left err ->
-         pure (IdeResponseFail
-                 (IdeError PluginError
-                           (T.pack $ "hare:findDefCmd" <> ": \"" <> err <> "\"")
-                           Null))
+getSymbolAtPoint :: Uri -> Position -> IdeM (IdeResponse (Maybe (Located Name)))
+getSymbolAtPoint file pos = do
+  mcm <- getCachedModule file
+  case mcm of
+    Nothing -> return $ nonExistentCacheErr "getSymbolAtPoint"
+    Just cm ->
+      return $ IdeResponseOk $ symbolFromTypecheckedModule (tcMod cm) =<< newPosToOld cm pos
+
+symbolFromTypecheckedModule :: TypecheckedModule -> Position -> Maybe (Located Name)
+symbolFromTypecheckedModule tc pos = do
+  pn@(L l _) <- locToRdrName (unPos pos) parsed
+  return $ L l $ rdrName2NamePure nameMap pn
+  where parsed = pm_parsed_source $ tm_parsed_module tc
+        nameMap = initRdrNameMap tc
 
+-- ---------------------------------------------------------------------
 
+invert :: (Ord k, Ord v) => Map.Map k v -> Map.Map v [k]
+invert m = Map.fromListWith (++) [(v,[k]) | (k,v) <- Map.toList m]
 
-getSymbolAtPoint :: Uri -> (Int,Int) -> IdeM (IdeResponse (Maybe (Located Name)))
-getSymbolAtPoint file (row,col) = do
-  cms <- getCachedModules
-  let mcm = Map.lookup file cms
+getReferencesInDoc :: Uri -> Position -> IdeM (IdeResponse [J.DocumentHighlight])
+getReferencesInDoc uri pos = runEitherT $ do
+  mcm <- lift $ getCachedModule uri
   case mcm of
-    Nothing ->
-      return $ IdeResponseFail
-        (IdeError PluginError
-                  (T.pack $ "hare:getSymbolAtPoint" <> ": \"" <> "module not loaded" <> "\"")
-                  Null)
+    Nothing -> hoistEither $ nonExistentCacheErr "getReferencesInDoc"
     Just cm -> do
       let tc = tcMod cm
           parsed = pm_parsed_source $ tm_parsed_module tc
           nameMap = initRdrNameMap tc
-      case locToRdrName (row, col) parsed of
-        Nothing -> return $ IdeResponseOk Nothing
-        Just pn@(L l _) -> pure $ IdeResponseOk (Just $ L l $ rdrName2NamePure nameMap pn)
+          invertedMap = invert nameMap
```

Comment:
You should consider putting this inside the cached module, so it gets inverted once, rather than on every use. Or would it get memoised?